File: DemoMMOUnity/Assets/Game/Net/Script/Net/CsProtocol/Protocol/ServerConfig/ServerConfigResponse.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using ZJYFrameWork.Collection.Reference;
using ZJYFrameWork.Net.Core;
using ZJYFrameWork.Net.CsProtocol.Buffer;
using ZJYFrameWork.Net.CsProtocol.Protocol.Bag;
using ZJYFrameWork.Net.CsProtocol.Protocol.Equipment;
using ZJYFrameWork.Spring.Utils;

namespace ZJYFrameWork.Net.CsProtocol.Protocol.ServerConfig
{
    /// <summary>
    /// 系统基础配置表相关
    /// </summary>
    public class ServerConfigResponse : Model, IPacket, IReference
    {
        public List<ItemBaseData> bagItemEntityList { get; set; } = default;

        /**
         * 装备基础相关 根据品节 卡等级 强化获取副属性条
        */
        public List<EquipmentConfigBaseData> equipmentConfigBaseDataList { get; set; } = default;

        /**
         * 圣遗物基础配置相关
         */
        public List<EquipmentBaseData> equipmentBaseDataList { get; set; } = default;

        /**
         * 圣遗物位置信息
         */
        public List<EquipmentPrimaryConfigBaseData> equipmentPrimaryConfigBaseDataList { get; set; } = default;

        /**
         * 圣遗物介绍
         */
        public List<EquipmentDesBaseData> equipmentDesBaseDataList { get; set; } = default;

        /**
         * 圣遗物位置名字
         */
        public List<EquipmentGrowthConfigBaseData> equipmentGrowthConfigBaseDataList { get; set; } = default;

        /**
         * 圣遗物副属性
         */
        public List<EquipmentGrowthViceConfigBaseData> equipmentGrowthViceConfigBaseDataList { get; set; } = default;

        public short ProtocolId()
        {
            return 1010;
        }

        public static ServerConfigResponse ValueOf()
        {
            var packet = ReferenceCache.Acquire<ServerConfigResponse>();
            packet.Clear();
            return packet;
        }

        public static ServerConfigResponse ValueOf(List<ItemBaseData> bagItemEntityList)
        {
            var packet = new ServerConfigResponse
            {
                bagItemEntityList = bagItemEntityList
            };
            return packet;
        }

        public void Clear()
        {
            bagItemEntityList = new List<ItemBaseData>();
            if (equipmentConfigBaseDataList != null)
            {
                equipmentConfigBaseDataList.Clear();
            }
            else
            {
                equipmentConfigBaseDataList = new List<EquipmentConfigBaseData>();
            }

            equipmentBaseDataList = new List<EquipmentBaseData>();
            equipmentPrimaryConfigBaseDataList = new List<EquipmentPrimaryConfigBaseData>();
            equipmentDesBaseDataList = new List<EquipmentDesBaseData>();
            equipmentGrowthConfigBaseDataList = new List<EquipmentGrowthConfigBaseData>();
            equipmentGrowthViceConfigBaseDataList = new List<EquipmentGrowthViceConfigBaseData>();
        }
    }

    public class ServerConfigResponseRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 1010;
        }

        public void Write(ByteBuffer buffer, IPacket packet)
        {
            var message = (ServerConfigResponse)packet;
            var messageWrite = new ServerMessageWrite(message.ProtocolId(), message);
            var json = JsonConvert.SerializeObject(messageWrite);
            buffer.WriteString(json);
        }

        public IPacket Read(ByteBuffer buffer)
        {
            var json = StringUtils.BytesToString(buffer.ToBytes());
            var packet = ServerConfigResponse.ValueOf();
            if (string.IsNullOrEmpty(json))
            {
                return packet;
            }

            var dict = JsonConvert.DeserializeObject<Dictionary<object, object>>(json);
            foreach (var (key, value) in dict)
            {
                var byteBuffers = ByteBuffer.ValueOf();
                var keyString = key.ToString();
                switch (keyString)
                {
                    case "bagItemEntityList":
                    {
                        var valueString = value.ToString();
                        var bagEntityList = JsonConvert.DeserializeObject<List<object>>(valueString);
                        int length = bagEntityList.Count;
                        packet.bagItemEntityList = new List<ItemBaseData>();
                        for (int i = 0; i < length; i++)
                        {
                            var entityObj = bagEntityList[i];
                            var packetData = ProtocolManager.GetProtocol(201);
                            byteBuffers.WriteString(entityObj.ToString());
                            var packetDataRead = (ItemBaseData)packetData.Read(byteBuffers);
                            packet.bagItemEntityList.Add(packetDataRead);
                            byteBuffers.Clear();
                        }
                    }
                        break;
                    case "equipmentBaseDataList":
                    {
                        var valueString = value.ToString();
                        var entityList = JsonConvert.DeserializeObject<List<object>>(valueString);
                        int length = entityList.Count;
                        for (int i = 0; i < length; i++)
                        {
                            var obj = entityList[i];
                            var packetData = ProtocolManager.GetProtocol(EquipmentBaseData.GetProtocolId());
                            byteBuffers.WriteString(obj.ToString());
                            var createPacket = (EquipmentBaseData)packetData.Read(byteBuffers);
                            packet.equipmentBaseDataList.Add(createPacket);
                            byteBuffers.Clear();
                        }
                    }
                        break;
                    case "equipmentConfigBaseDataList":
                    {
                        var valueString = value.ToString();
                        var entityList = JsonConvert.DeserializeObject<List<object>>(valueString);
                        int length = entityList.Count;
                        for (int i = 0; i < length; i++)
                        {
                            var obj = entityList[i];
                            var packetData = ProtocolManager.GetProtocol(EquipmentConfigBaseData.GetProtocolId());
                            byteBuffers.WriteString(obj.ToString());
                            var createPacket = (EquipmentConfigBaseData)packetData.Read(byteBuffers);
                            packet.equipmentConfigBaseDataList.Add(createPacket);
                            byteBuffers.Clear();
                        }
                    }
                        break;
                    case "equipmentPrimaryConfigBaseDataList":
                    {
                        var valueString = value.ToString();
                        var entityList = JsonConvert.DeserializeObject<List<object>>(valueString);
                        int length = entityList.Count;
                        for (int i = 0; i < length; i++)
                        {
                            var obj = entityList[i];
                            var packetData = ProtocolManager.GetProtocol(EquipmentPrimaryConfigBaseData.GetProtocolId());
                            byteBuffers.WriteString(obj.ToString());
                            var createPacket = (EquipmentPrimaryConfigBaseData)packetData.Read(byteBuffers);
                            packet.equipmentPrimaryConfigBaseDataList.Add(createPacket);
                            byteBuffers.Clear();
                        }
                    }
                        break;
                    case "equipmentDesBaseDataList":
                    {
                        var valueString = value.ToString();
                        var entityList = JsonConvert.DeserializeObject<List<object>>(valueString);
                        int length = entityList.Count;
                        for (int i = 0; i < length; i++)
                        {
                            var obj = entityList[i];
                            var packetData = ProtocolManager.GetProtocol(EquipmentDesBaseData.GetProtocolId());
                            byteBuffers.WriteString(obj.ToString());
                            var createPacket = (EquipmentDesBaseData)packetData.Read(byteBuffers);
                            packet.equipmentDesBaseDataList.Add(createPacket);
                            byteBuffers.Clear();
                        }
                    }
                        break;
                    case "equipmentGrowthConfigBaseDataList":
                    {
                        var valueString = value.ToString();
                        var entityList = JsonConvert.DeserializeObject<List<object>>(valueString);
                        int length = entityList.Count;
                        for (int i = 0; i < length; i++)
                        {
                            var obj = entityList[i];
                            var packetData = ProtocolManager.GetProtocol(EquipmentGrowthConfigBaseData.GetProtocolId());
                            byteBuffers.WriteString(obj.ToString());
                            var createPacket = (EquipmentGrowthConfigBaseData)packetData.Read(byteBuffers);
                            packet.equipmentGrowthConfigBaseDataList.Add(createPacket);
                            byteBuffers.Clear();
                        }
                    }
                        break;
                    case "equipmentGrowthViceConfigBaseDataList":
                    {
                        var valueString = value.ToString();
                        var entityList = JsonConvert.DeserializeObject<List<object>>(valueString);
                        int length = entityList.Count;
                        for (int i = 0; i < length; i++)
                        {
                            var obj = entityList[i];
                            var packetData = ProtocolManager.GetProtocol(EquipmentGrowthViceConfigBaseData.GetProtocolId());
                            byteBuffers.WriteString(obj.ToString());
                            var createPacket = (EquipmentGrowthViceConfigBaseData)packetData.Read(byteBuffers);
                            packet.equipmentGrowthViceConfigBaseDataList.Add(createPacket);
                            byteBuffers.Clear();
                        }
                    }
                        break;
                }
            }

            return packet;
        }
    }
}

File: UnityMMOClient/Lua/application/app/net/webSocket/luaProtocol/serverConfig/serverConfigResponse.lua
---@class ServerConfigResponse
local ServerConfigResponse = class("ServerConfigResponse")
function ServerConfigResponse:ctor()
    --- 背包基础类list
    ---@type  table<number,ItemBaseData>
    self.bagItemEntityList = {}
    --- 圣遗物基础配置相关
    ---@type  table<number,EquipmentBaseData>
    self.equipmentBaseDataList = {}
    --- 装备基础相关根据品节卡等级强化获取副属性条
    ---@type  table<number,EquipmentConfigBaseData>
    self.equipmentConfigBaseDataList = {}
    --- 圣遗物介绍
    ---@type  table<number,EquipmentDesBaseData>
    self.equipmentDesBaseDataList = {}
    --- 圣遗物位置名字
    ---@type  table<number,EquipmentGrowthConfigBaseData>
    self.equipmentGrowthConfigBaseDataList = {}
    --- 圣遗物副属性
    ---@type  table<number,EquipmentGrowthViceConfigBaseData>
    self.equipmentGrowthViceConfigBaseDataList = {}
    --- 圣遗物位置信息
    ---@type  table<number,EquipmentPrimaryConfigBaseData>
    self.equipmentPrimaryConfigBaseDataList = {}
    --- 武器配置
    ---@type  table<number,WeaponsConfigData>
    self.weaponsConfigDataList = {}
end

---@param bagItemEntityList table<number,ItemBaseData> 背包基础类list
---@param equipmentBaseDataList table<number,EquipmentBaseData> 圣遗物基础配置相关
---@param equipmentConfigBaseDataList table<number,EquipmentConfigBaseData> 装备基础相关根据品节卡等级强化获取副属性条
---@param equipmentDesBaseDataList table<number,EquipmentDesBaseData> 圣遗物介绍
---@param equipmentGrowthConfigBaseDataList table<number,EquipmentGrowthConfigBaseData> 圣遗物位置名字
---@param equipmentGrowthViceConfigBaseDataList table<number,EquipmentGrowthViceConfigBaseData> 圣遗物副属性
---@param equipmentPrimaryConfigBaseDataList table<number,EquipmentPrimaryConfigBaseData> 圣遗物位置信息
---@param weaponsConfigDataList table<number,WeaponsConfigData> 武器配置
---@return ServerConfigResponse
function ServerConfigResponse:new(bagItemEntityList, equipmentBaseDataList, equipmentConfigBaseDataList, equipmentDesBaseDataList, equipmentGrowthConfigBaseDataList, equipmentGrowthViceConfigBaseDataList, equipmentPrimaryConfigBaseDataList, weaponsConfigDataList)
    self.bagItemEntityList = bagItemEntityList --- java.util.List<com.gameServer.common.protocol.serverConfig.ItemBaseData>
    self.equipmentBaseDataList = equipmentBaseDataList --- java.util.List<com.gameServer.common.protocol.equipment.base.EquipmentBaseData>
    self.equipmentConfigBaseDataList = equipmentConfigBaseDataList --- java.util.List<com.gameServer.common.protocol.equipment.base.EquipmentConfigBaseData>
    self.equipmentDesBaseDataList = equipmentDesBaseDataList --- java.util.List<com.gameServer.common.protocol.equipment.base.EquipmentDesBaseData>
    self.equipmentGrowthConfigBaseDataList = equipmentGrowthConfigBaseDataList --- java.util.List<com.gameServer.common.protocol.equipment.base.EquipmentGrowthConfigBaseData>
    self.equipmentGrowthViceConfigBaseDataList = equipmentGrowthViceConfigBaseDataList --- java.util.List<com.gameServer.common.protocol.equipment.base.EquipmentGrowthViceConfigBaseData>
    self.equipmentPrimaryConfigBaseDataList = equipmentPrimaryConfigBaseDataList --- java.util.List<com.gameServer.common.protocol.equipment.base.EquipmentPrimaryConfigBaseData>
    self.weaponsConfigDataList = weaponsConfigDataList --- java.util.List<com.gameServer.common.protocol.weapon.WeaponsConfigData>
    return self
end

---@return number
function ServerConfigResponse:protocolId()
    return 1010
end

---@return string
function ServerConfigResponse:write()
    local message = {
        protocolId = self:protocolId(),
        packet = {
            bagItemEntityList = self.bagItemEntityList,
            equipmentBaseDataList = self.equipmentBaseDataList,
            equipmentConfigBaseDataList = self.equipmentConfigBaseDataList,
            equipmentDesBaseDataList = self.equipmentDesBaseDataList,
            equipmentGrowthConfigBaseDataList = self.equipmentGrowthConfigBaseDataList,
            equipmentGrowthViceConfigBaseDataList = self.equipmentGrowthViceConfigBaseDataList,
            equipmentPrimaryConfigBaseDataList = self.equipmentPrimaryConfigBaseDataList,
            weaponsConfigDataList = self.weaponsConfigDataList
        }
    }
    local jsonStr = JSON.encode(message)
    return jsonStr
end

---@return ServerConfigResponse
function ServerConfigResponse:read(data)
    local bagItemEntityList = {}
    for index, value in ipairs(data.bagItemEntityList) do
        local bagItemEntityListPacket = ItemBaseData()
        local packetData = bagItemEntityListPacket:read(value)
        table.insert(bagItemEntityList,packetData)
    end
    local equipmentBaseDataList = {}
    for index, value in ipairs(data.equipmentBaseDataList) do
        local equipmentBaseDataListPacket = EquipmentBaseData()
        local packetData = equipmentBaseDataListPacket:read(value)
        table.insert(equipmentBaseDataList,packetData)
    end
    local equipmentConfigBaseDataList = {}
    for index, value in ipairs(data.equipmentConfigBaseDataList) do
        local equipmentConfigBaseDataListPacket = EquipmentConfigBaseData()
        local packetData = equipmentConfigBaseDataListPacket:read(value)
        table.insert(equipmentConfigBaseDataList,packetData)
    end
    local equipmentDesBaseDataList = {}
    for index, value in ipairs(data.equipmentDesBaseDataList) do
        local equipmentDesBaseDataListPacket = EquipmentDesBaseData()
        local packetData = equipmentDesBaseDataListPacket:read(value)
        table.insert(equipmentDesBaseDataList,packetData)
    end
    local equipmentGrowthConfigBaseDataList = {}
    for index, value in ipairs(data.equipmentGrowthConfigBaseDataList) do
        local equipmentGrowthConfigBaseDataListPacket = EquipmentGrowthConfigBaseData()
        local packetData = equipmentGrowthConfigBaseDataListPacket:read(value)
        table.insert(equipmentGrowthConfigBaseDataList,packetData)
    end
    local equipmentGrowthViceConfigBaseDataList = {}
    for index, value in ipairs(data.equipmentGrowthViceConfigBaseDataList) do
        local equipmentGrowthViceConfigBaseDataListPacket = EquipmentGrowthViceConfigBaseData()
        local packetData = equipmentGrowthViceConfigBaseDataListPacket:read(value)
        table.insert(equipmentGrowthViceConfigBaseDataList,packetData)
    end
    local equipmentPrimaryConfigBaseDataList = {}
    for index, value in ipairs(data.equipmentPrimaryConfigBaseDataList) do
        local equipmentPrimaryConfigBaseDataListPacket = EquipmentPrimaryConfigBaseData()
        local packetData = equipmentPrimaryConfigBaseDataListPacket:read(value)
        table.insert(equipmentPrimaryConfigBaseDataList,packetData)
    end
    local weaponsConfigDataList = {}
    for index, value in ipairs(data.weaponsConfigDataList) do
        local weaponsConfigDataListPacket = WeaponsConfigData()
        local packetData = weaponsConfigDataListPacket:read(value)
        table.insert(weaponsConfigDataList,packetData)
    end

    local packet = self:new(
            bagItemEntityList,
            equipmentBaseDataList,
            equipmentConfigBaseDataList,
            equipmentDesBaseDataList,
            equipmentGrowthConfigBaseDataList,
            equipmentGrowthViceConfigBaseDataList,
            equipmentPrimaryConfigBaseDataList,
            weaponsConfigDataList)
    return packet
end

--- 背包基础类list
---@type  table<number,ItemBaseData> 背包基础类list
function ServerConfigResponse:getBagItemEntityList()
    return self.bagItemEntityList
end
--- 圣遗物基础配置相关
---@type  table<number,EquipmentBaseData> 圣遗物基础配置相关
function ServerConfigResponse:getEquipmentBaseDataList()
    return self.equipmentBaseDataList
end
--- 装备基础相关根据品节卡等级强化获取副属性条
---@type  table<number,EquipmentConfigBaseData> 装备基础相关根据品节卡等级强化获取副属性条
function ServerConfigResponse:getEquipmentConfigBaseDataList()
    return self.equipmentConfigBaseDataList
end
--- 圣遗物介绍
---@type  table<number,EquipmentDesBaseData> 圣遗物介绍
function ServerConfigResponse:getEquipmentDesBaseDataList()
    return self.equipmentDesBaseDataList
end
--- 圣遗物位置名字
---@type  table<number,EquipmentGrowthConfigBaseData> 圣遗物位置名字
function ServerConfigResponse:getEquipmentGrowthConfigBaseDataList()
    return self.equipmentGrowthConfigBaseDataList
end
--- 圣遗物副属性
---@type  table<number,EquipmentGrowthViceConfigBaseData> 圣遗物副属性
function ServerConfigResponse:getEquipmentGrowthViceConfigBaseDataList()
    return self.equipmentGrowthViceConfigBaseDataList
end
--- 圣遗物位置信息
---@type  table<number,EquipmentPrimaryConfigBaseData> 圣遗物位置信息
function ServerConfigResponse:getEquipmentPrimaryConfigBaseDataList()
    return self.equipmentPrimaryConfigBaseDataList
end
--- 武器配置
---@type  table<number,WeaponsConfigData> 武器配置
function ServerConfigResponse:getWeaponsConfigDataList()
    return self.weaponsConfigDataList
end


return ServerConfigResponse
